WIP Control flow

rocky opened the pull request:
Probably need lingo to tell byte-recalc-examples to use optimization. 

pdf sectioning in evince is weird, and I don't know why

rocky committed:
Start documenting format. Add goto instruction

rocky committed:
Start documenting format. Add goto instruction

rocky committed:
Minor corrections

rocky committed:
Minor corrections

rocky committed:
More corrections..

Add goto and goto-if-nil in opcode table

rocky committed:
More corrections..

Add goto and goto-if-nil in opcode table

rocky committed:
Merge branch 'control-flow' of github.com:rocky/elisp-lap into control-flow

rocky committed:
Another pass at control flow; Start adding instruction sizes

pipcet commented on instructions.texi:
```diff
@@ -206,23 +301,185 @@ PC  Byte  Instruction
+@kindex goto-if-not-nil
+@table @strong
+@item Implements:
+Jump to label given in the 16-bit operand if TOS is not nil
```
I'd explicitly point out that TOS is always popped

rocky committed:
What's up with goto..-else-pop...

Be more description about jump instructions and fill in the missing
example. More regular use of constants in examples.

rocky committed:
Note obsolete instruction scan-buffer

rocky commented:
Ok. How about now? (Feel free just to add changes on top of this if that's easier.)

rocky committed:
Fix example name

pipcet commented on instructions.texi:
```diff
@@ -1,15 +1,110 @@
+number of opcodes in a class encodes an index into the opcode. We
+generally call this an ``Argument-encoding'' instruction. In both the
+display of the opcode as well we'll include that particular opcode
+variant in subscripts.
```
It's not clear to me here what the "both" was meant to refer to.

rocky committed:
Clarify opcode index

rocky commented:
Thanks - have clarified what I meant by "both".

pipcet reviewed (approved):
This is excellent, I like the way things are going.

rocky commented:
Thanks for the kind words. And thanks for the detailed reading and feedback.

I am overwhelmed by the amount of work yet to do. 

At some point we need to make another pass over the opcode table to fix the formatting yet again. 

My current thought there is to add line breaks in the opcode name, and possibly reduce the size of the font, and reduce the description (analogous to what is in bytecode.c as the description), but include a link to the long description. But also we should add a field for instruction size.

pipcet commented:
That sounds like a good plan, though I'd kind of thought we'd leave the reformatting for a while and focus on adding content. I'm happy either way, just let me know what's your priority.

rocky commented:
In general I don't have a predisposed preference when there is so much to do. I'm cool with deferring the reformatting.  

(There seem to be a whole slew of texinfo related weirdnesses too that just have to do with coaxing texinfo. In evince I am getting weird expand lists on the left. Things like references to links don't just to the right place example and this also happens in the table of contents and index. In info format though this generally works.)

In general my work plan is to pick out the hardest thing to do that I am up for. I was traveling when I started the control flow stuff, and in traveling which is full of interruptions I needed work that could be interruptible. 

In terms of content, we still need a section on bytecode files. I think there should also be a section that goes through working out a small example LAP program and how that works. 

Also I think we should start to document the _types_ of optimization that is done -- and not, e.g. tail recursion. 

But as for instructions. Declare a section and work on it. When I work on instructions next, I'll do the Lisp function instructions as that seems to be in between two sections that are currently done.
